### PR TITLE
Add additionalPrinterColumns to CAPZ resources

### LINKED ISSUE #4376: Add Kubebuilder `printcolumns` to CAPZ resources
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
- This ask is evolving out of the comment at https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4371#issuecomment-1853191449.

- We should be adding more `print columns` to `azuremanagedcontrolplanes.infrastructure.cluster.x-k8s.io` to display the resource's state. As of now, we only display the resource name and its age.
```
❯ kg azuremanagedcontrolplanes.infrastructure.cluster.x-k8s.io
NAME        AGE
aks-24401   3m49s
```

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
- There are many other resource types in CAPZ that need more printcolumns added to them. We can either make this issue an epic or have individual issues connected to it. Either works.

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by prashantrewar**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4376 
**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Add additional Printer Columns like "Ready", "Severity", "Reason" and "Message" to CAPZ resources
```

### CODE REVIEW

File: config/crd/bases/infrastructure.cluster.x-k8s.io_azuremanagedcontrolplanes.yaml
Comment (CecileRobertMichon) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_azuremanagedcontrolplanes.yaml:21]: this should be auto generated by adding kubebuilder annotations on the type definitions


See https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/azurecluster_types.go#L80 for an example

Comment (mboersma) [reply config/crd/bases/infrastructure.cluster.x-k8s.io_azuremanagedcontrolplanes.yaml:21]: The file edited here is actually generated, so we shouldn't make changes to it.

Instead, add a [kubebuilder annotation](https://book.kubebuilder.io/reference/markers.html) similar to what @CecileRobertMichon linked to above, and then run `make generate`.

Comment (prashantrewar) [reply config/crd/bases/infrastructure.cluster.x-k8s.io_azuremanagedcontrolplanes.yaml:21]: done!, could you please a look

**Comment by prashantrewar** [comment]
@mboersma, could you please rerun the tests

**Comment by willie-yao** [comment]
One comment: Make sure to add a release note and add `/kind feature` to the PR description instead of `kind/feature`

**Comment by prashantrewar** [comment]
> One comment: Make sure to add a release note and add `/kind feature` to the PR description instead of `kind/feature`

Done! @willie-yao

File: api/v1beta1/azureclusteridentity_types.go
Comment (CecileRobertMichon) [inline api/v1beta1/azureclusteridentity_types.go:80]: Where is this label getting set? I don't see it when trying out this PR locally

```
k get azureclusteridentities.infrastructure.cluster.x-k8s.io
NAME               CLUSTER   READY   SEVERITY   REASON   TYPE               AGE
cluster-identity                                         ServicePrincipal   9m27s
```

File: api/v1beta1/azureclusteridentity_types.go
Comment (nojnhuh) [inline api/v1beta1/azureclusteridentity_types.go:84]: Are these fields meaningful for an AzureClusterIdentity? I don't see where we'd ever set `status.conditions` at all for one.

File: api/v1beta1/azuremanagedmachinepool_types.go
Comment (nojnhuh) [inline api/v1beta1/azuremanagedmachinepool_types.go:491]: Was this meant for AzureClusterIdentity?

File: api/v1beta1/azuremanagedcluster_types.go
Comment (nojnhuh) [inline api/v1beta1/azuremanagedcluster_types.go:45]: I'd vote to remove this one since AzureManagedCluster has no `spec.type` field.

File: api/v1beta1/azureclusteridentity_types.go
Comment (nojnhuh) [reply api/v1beta1/azureclusteridentity_types.go:80]: I don't think it ever does. And I'm not sure it would make sense to add it since an AzureClusterIdentity can be shared across multiple clusters, so I'd vote to remove this.

File: api/v1beta1/azuremanagedcluster_types.go
Comment (prashantrewar) [reply api/v1beta1/azuremanagedcluster_types.go:45]: So should I remove only the 'type' field or should I remove all fields like 'ready', 'severity', 'reason', and 'message'?

File: api/v1beta1/azuremanagedmachinepool_types.go
Comment (prashantrewar) [reply api/v1beta1/azuremanagedmachinepool_types.go:491]: I apologize, I didn't see that. No, it is for AzureManagedMachinePool.

File: api/v1beta1/azureclusteridentity_types.go
Comment (prashantrewar) [reply api/v1beta1/azureclusteridentity_types.go:84]: So should I remove all fields?

Comment (nojnhuh) [reply api/v1beta1/azureclusteridentity_types.go:84]: Yes, I think we can remove all of these. The `spec.type` one currently defined on AzureManagedCluster could still be useful though.

File: api/v1beta1/azuremanagedmachinepool_types.go
Comment (nojnhuh) [reply api/v1beta1/azuremanagedmachinepool_types.go:491]: I think we should remove this then since AzureManagedMachinePool doesn't have a `spec.type` field.

File: api/v1beta1/azuremanagedcluster_types.go
Comment (nojnhuh) [reply api/v1beta1/azuremanagedcluster_types.go:45]: I was referring only to the `spec.type` one.

**Comment by prashantrewar** [comment]
Hey @nojnhuh, @CecileRobertMichon, I updated the changes, could you please take a look?

**Comment by prashantrewar** [comment]
Hey @nojnhuh, @CecileRobertMichon, I updated the changes, could you please take a look?

**Comment by nawazkh** [comment]
Not sure how I lost track of this PR. I am sorry.
Thank you for putting in so much effort @prashantrewar ❤️ 
This PR has been vetted already, but holding it for a bit before it gets lgtm-ed and approved. 
/hold

**Review by nawazkh — APPROVED** [review]
Great work, thanks for putting it together!
/unhold

**Comment by mboersma** [comment]
/hold

I'm just testing this locally.

File: api/v1beta1/azuremanagedcluster_types.go
Comment (mboersma) [inline api/v1beta1/azuremanagedcluster_types.go:41]: This doesn't seem to work on an `AzureManagedCluster` because its status field doesn't apparently have `conditions`:

```shell
k get azuremanagedcluster,azuremanagedcontrolplane
NAME                                                            CLUSTER     READY   SEVERITY   REASON   AGE
azuremanagedcluster.infrastructure.cluster.x-k8s.io/aks-17271   aks-17271                               13m

NAME                                                                 CLUSTER     READY   SEVERITY   REASON   AGE
azuremanagedcontrolplane.infrastructure.cluster.x-k8s.io/aks-17271   aks-17271   True                        13m

% k get azuremanagedcluster,azuremanagedcontrolplane -o yaml
apiVersion: v1
items:
- apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
  kind: AzureManagedCluster
  <...>
  status:
    ready: true
- apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
  kind: AzureManagedControlPlane
  <...>
  status:
    conditions:
    - lastTransitionTime: "2024-01-24T00:00:19Z"
      status: "True"
      type: Ready
    - lastTransitionTime: "2024-01-24T00:00:19Z"
      status: "True"
      type: ManagedClusterRunning
    - lastTransitionTime: "2024-01-23T23:52:27Z"
      status: "True"
      type: ResourceGroupReady
    - lastTransitionTime: "2024-01-23T23:53:28Z"
      status: "True"
      type: SubnetsReady
    - lastTransitionTime: "2024-01-23T23:52:58Z"
      status: "True"
      type: VNetReady
    initialized: true
    ready: true
    version: v1.28.3
```

Comment (nawazkh) [reply api/v1beta1/azuremanagedcluster_types.go:41]: :/

Comment (prashantrewar) [reply api/v1beta1/azuremanagedcluster_types.go:41]: I apologize I didn't notice that so should I remove all fields like 'Ready', 'Severity', and 'Reason'?

Comment (mboersma) [reply api/v1beta1/azuremanagedcluster_types.go:41]: > so should I remove all fields ?

I think we have to, since conditions aren't supported here. We can keep "Ready" as just a boolean field though.

**Comment by prashantrewar** [comment]
Hey @mboersma and @nawazkh, I updated the changes. Please take a look and suggest changes if required.

File: api/v1beta1/azuremanagedcluster_types.go
Comment (mboersma) [inline api/v1beta1/azuremanagedcluster_types.go:41]: I think this still won't work, since the `JSONPath` query refers to `conditions`.

Comment (mboersma) [inline api/v1beta1/azuremanagedcluster_types.go:41]:
```suggestion
// +kubebuilder:printcolumn:name="Ready",type="string",JSONPath=".status.ready"
```

**Review by mboersma — CHANGES_REQUESTED** [review]
I think this PR is ready to go if you can change the AzureManagedCluster `ready` field. I tested with that change and get this:

```sh
NAME                                                           CLUSTER    READY   AGE
azuremanagedcluster.infrastructure.cluster.x-k8s.io/aks-1777   aks-1777   true    61m
```

**Comment by prashantrewar** [comment]
Done! @mboersma, I update the changes.

Thank you so much for guiding me.

**Comment by mboersma** [comment]
Thanks @prashantrewar. It looks like a random file `exp/controllers/mocks/gomock_reflect_3841401974/prog.go` got added to the commit. Otherwise, lgtm!

**Comment by prashantrewar** [comment]
> It looks like a random file `exp/controllers/mocks/gomock_reflect_3841401974/prog.go` got added to the commit. 

I think it is an auto-generated file, I did't add this. What should I do Now for this?

**Comment by nawazkh** [comment]
> @mboersma I think it is an auto-generated file, I did't add this. What should I do Now for this?

Try deleting that file, commit, squash commits and force push. That could help.

**Comment by prashantrewar** [comment]
> > @mboersma I think it is an auto-generated file, I did't add this. What should I do Now for this?
> 
> Try deleting that file, commit, squash commits and force push. That could help.

Done!

**Review by mboersma — APPROVED** [review]

**Comment by nawazkh** [comment]
Thank you @prashantrewar for all the work, and @mboersma for helping out 
/lgtm

**Pull request merged**
